dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1390 @@
+                                  floating[:,:,:,:] mgradient, int[:,:,:] smask,
+                                  int[:,:,:] mmask, double smin, double sdelta,
+                                  double mmin, double mdelta, int nbins,
+                                  int padding, double[:,:,:] grad_pdf):
+    r''' Gradient of the joint PDF w.r.t. transform parameters theta
+
+    Computes the vector of partial derivatives of the joint histogram w.r.t.
+    each transformation parameter. The transformation itself is not necessary
+    to compute the gradient, but only its Jacobian.
+
+    Parameters
+    ----------
+    theta: array, shape (n,)
+        parameters of the transformation to compute the gradient from
+    transform: instance of Transform
+        the transformation with respect to whose parameters the gradient
+        must be computed
+    static: array, shape (S, R, C)
+        static image
+    moving: array, shape (S, R, C)
+        moving image
+    grid_to_space: array, shape (4, 4)
+        the grid-to-space transform associated to images static and moving
+        (we assume that both images have already been sampled at a common grid)
+    mgradient: array, shape (S, R, C, 3)
+        the gradient of the moving image
+    smask: array, shape (S, R, C)
+        mask of static object being registered (a binary array with 1's inside
+        the object of interest and 0's along the background)
+    mmask: array, shape (S, R, C)
+        mask of moving object being registered (a binary array with 1's inside
+        the object of interest and 0's along the background)
+    smin: float
+        the minimum observed intensity associated to the static image, which
+        was used to define the joint PDF
+    sdelta: float
+        bin size associated to the intensities of the static image
+    mmin: float
+        the minimum observed intensity associated to the moving image, which
+        was used to define the joint PDF
+    mdelta: float
+        bin size associated to the intensities of the moving image
+    nbins: int
+        number of histogram bins
+    padding: int
+        number of bins used as padding (the total bins used for padding at both
+        sides of the histogram is actually 2*padding)
+    grad_pdf: array, shape (nbins, nbins, len(theta))
+        the array to write the gradient to
+    '''
+    cdef:
+        int nslices = static.shape[0]
+        int nrows = static.shape[1]
+        int ncols = static.shape[2]
+        int n = theta.shape[0]
+        int offset, valid_points, constant_jacobian=0
+        cnp.npy_intp l, k, i, j, r, c
+        double rn, cn
+        double val, spline_arg
+        double[:,:] J = np.ndarray(shape=(3, n), dtype=np.float64)
+        double[:] prod = np.ndarray(shape=(n,), dtype=np.float64)
+        double[:] x = np.ndarray(shape=(3,), dtype=np.float64)
+
+    grad_pdf[...] = 0
+    with nogil:
+        valid_points = 0
+        for k in range(nslices):
```

Comment:
Yes, it is an old bad habit too ='(, since I used to code everything in c++, for me the data buffer was organized as consecutive slices (each representing a 2D image). By indexing as (k, i, j) I had the same index convention for 3D slices and 2D images. If I used "i" to identify the slice (with (i, j, k) ), then the second and third indices would be "j"(row) and "k"(column) making it inconsistent with the "i"(row) and "j"(column) indexing for 2D images. This kind of things were extremely confusing for me at the beginning, now it's just very confusing, hopefully at some point it will be just confusing...